pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -691,6 +698,46 @@ func (ssc *defaultStatefulSetControl) rollingUpdateStatefulsetPods(
 	return status, nil
 }
 
+func (ssc *defaultStatefulSetControl) handleSpecifiedDeletedPods(
+	set *appsv1beta1.StatefulSet,
+	status *appsv1beta1.StatefulSetStatus,
+	currentRevision *apps.ControllerRevision,
+	updateRevision *apps.ControllerRevision,
+	replicas []*v1.Pod,
+	maxUnavailable int,
+	unavailablePods sets.String) (sets.String, error) {
+	specifiedDeletedPods := sets.NewString()
+	for target := len(replicas) - 1; target >= 0; target-- {
+		if replicas[target] == nil || !specifieddelete.IsSpecifiedDelete(replicas[target]) {
+			continue
+		}
+		// the unavailable pods count exceed the maxUnavailable and the target is available, so we can't process it,
+		// why skip here rather than return?
+		// case: pod 0 ready, pod1 unready, pod 2 unready, pod3 ready, pod4 ready
+		// when maxUnavailable = 3, pod4 with specified deleted will be deleted but pod3 can't
+		// pod 2 and pod 1 can be deleted because they were unavailable
+		if len(unavailablePods) >= maxUnavailable && !unavailablePods.Has(replicas[target].Name) {
+			klog.V(4).InfoS("StatefulSet was waiting for unavailable Pods to update, blocked pod",
+				"statefulSet", klog.KObj(set), "unavailablePods", unavailablePods.List(), "blockedPod", klog.KObj(replicas[target]))
+			continue
+		}
+
+		specifiedDeletedPods.Insert(replicas[target].Name)
+		if _, err := ssc.deletePod(set, replicas[target]); err != nil {
```

Comment:
before update replicas in status, we should check whether the pod is really deleted, not just blocked in the predelete hook

Reply:
Now I will decrease revision if the pod is actual deleting.